drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1508 @@
+  enum rn2xx3_type_e model; /* Transceiver model */
+  enum rn2xx3_mod_e mod;    /* Modulation mode */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static ssize_t rn2xx3_write(FAR struct file *filep, FAR const char *buffer,
+                            size_t buflen);
+static ssize_t rn2xx3_read(FAR struct file *filep, FAR char *buffer,
+                           size_t buflen);
+static int rn2xx3_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_rn2xx3fops =
+{
+  .open = NULL,
+  .close = NULL,
+  .read = rn2xx3_read,
+  .write = rn2xx3_write,
+  .ioctl = rn2xx3_ioctl,
+};
+
+/* Modulation types */
+
+static const char *MODULATIONS[] =
+{
+  "lora", /* RN2XX3_MOD_LORA */
+  "fsk",  /* RN2XX3_MOD_FSK */
+};
+
+/* Max packet sizes for each modulation type. */
+
+static const uint8_t MAX_PKTSIZE[] =
+{
+  255, /* RN2XX3_MOD_LORA */
+  64,  /* RN2XX3_MOD_FSK */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: radio_flush
+ *
+ * Description:
+ *   Flushes all characters in the receive buffer from the RN2xx3 radio.
+ *   Returns the number of bytes read.
+ *
+ ****************************************************************************/
+
+static int radio_flush(FAR struct rn2xx3_dev_s *priv)
+{
+  int err;
+  int to_read;
+  char buf[10];
+
+  err = file_ioctl(&priv->uart, FIONREAD, &to_read);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  while (to_read > 0)
+    {
+      err = file_read(&priv->uart, buf, sizeof(buf));
+      if (err < 0)
+        {
+          return err;
+        }
+
+      to_read -= err;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: read_line
+ *
+ * Description:
+ *   Reads a whole line of response from the RN2xx3 radio. Returns the number
+ *   of bytes read.
+ *
+ ****************************************************************************/
+
+static ssize_t read_line(FAR struct rn2xx3_dev_s *priv, FAR char *buf,
+                         size_t nbytes)
+{
+  ssize_t length = 0;
+  ssize_t i = 0;
+
+  for (; i < nbytes; )
+    {
+      length = file_read(&priv->uart, &buf[i], nbytes - i);
+
+      if (length < 0)
+        {
+          return length;
+        }
+
+      i += length;
+
+      if (buf[i] == '\n' || (buf[i - 1] == '\n' && i > 0))
+        {
+          break;
+        }
+    }
+
+  return i + 1; /* Number of bytes read */
```

Comment:
Why i+1 here? You don's seem to be writing extra bytes to `buf`

Reply:
`i` counts from 0 up to `nbytes`. So when the loop is finished, `i` will have the value of `nbytes - 1`

Reply:
If say `nbytes` is 10, then `file_read(..., 10)` may return, let's say 7 - you read 7 bytes into buf. So `i` will now also be 7. Now when next condition is true, you will return `7 + 1` which is 8, and you read 7 bytes.

Please correct me if I'm wrong.

Reply:
You're right, I was thinking about this incorrectly!